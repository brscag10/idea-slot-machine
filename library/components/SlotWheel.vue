<template>
  <section
    class="SlotWheel"
    v-bind:class="{ 'SlotWheel--set': hasSpun && !isSpinning, 'SlotWheel--tick': hasTicked }"
    v-bind:style="{ 'animation-duration': `${tickDuration}ms` }"
    @animationend="_clearTick"
  >
    <div class="SlotWheel__value">{{ displayValue }}</div>
    <h2 class="SlotWheel__name">{{ name }}</h2>
  </section>
</template>


<script lang="ts">
import Component from "vue-class-component";
import ShuffleQueue from "../classes/ShuffleQueue";
import Vue from "vue";

import config from "app.config.yml";
import randomIntFactory from "../functions/randomIntFactory";

@Component({
  props: {
    name: String,
    options: Array
  }
})
/** 
 * the SlotWheel, once spun, triggers at a regular interval, taking care to not
 * repeat the same thing twice.
 */
export default class SlotWheel extends Vue {
  displayValue: string = "???";
  hasSpun: boolean = false;
  hasTicked: boolean = false;
  isSpinning: boolean = false;
  queue: ShuffleQueue<string> = new ShuffleQueue<string>(this.options);
  tickDuration: number = 0;

  startingSlotVelocityMS: Range<number> = config.startingSlotVelocityMS;
  finalTickInterval: Range<number> = config.finalTickInterval;

  /**
   * randomizes the final tick duration
   * 
   * @returns {number} the duration, in MS
   */
  randomFinalTickInterval: () => number = () => randomIntFactory(
    this.finalTickInterval.max,
    this.finalTickInterval.min
  )();

  /**
   * randomizes the initial tick velocity
   * 
   * @returns {number} the velocity, in MS
   */
  randomStartingVelocity: () => number = () => randomIntFactory(
    this.startingSlotVelocityMS.max,
    this.startingSlotVelocityMS.min
  )();

  /**
   * spin() starts spinning the slot wheel, once called.
   * 
   * @returns {Promise<void>} a promise that resolves once the wheel has stopped spinning
   */
  spin() {
    this.hasSpun = true;
    this.isSpinning = true;

    const startingVelocity = this.randomStartingVelocity();
    const finalTickInterval = this.randomFinalTickInterval();

    let nextTimeout = startingVelocity;

    return new Promise(resolve => {
      this._tick({ finalTickInterval, nextTimeout }, resolve);
    });
  }

  /**
   * _tick uses the configured frictionCoefficient to determine whether to
   * calculate another PhysicsFrame or stop the spinner
   * 
   * @param {function} resolve callback to invoke if this is the last tick on the wheel
   * @returns {void}
   */
  _tick({ finalTickInterval, nextTimeout }: PhysicsFrame, resolve) {
    this.displayValue = this.queue.pick();

    nextTimeout *= config.frictionCoeffecient;

    if (nextTimeout > finalTickInterval) {
      this.isSpinning = false;
      resolve();
    } else {
      this.tickDuration = nextTimeout / 2;
      this.hasTicked = true;

      setTimeout(() => this._tick({ finalTickInterval, nextTimeout }, resolve), nextTimeout);
    }
  }

  /**
   * clears the `hasTicked` property, which preps the SlotWheel for its next animation
   * 
   * @returns {void}
   */
  _clearTick() {
    this.hasTicked = false;
  }
}
</script>


<style scoped>
.SlotWheel {
  height: calc(var(--title-size) * 6);
  padding: var(--vert-padding) var(--horiz-padding);

  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;

  overflow: hidden;
  border-radius: calc(var(--horiz-padding) / 2);
  background: var(--background-color);

  transition: background-color 350ms ease-out;

  will-change: background-color;
}

.SlotWheel--tick {
  animation-name: tick;
}

@keyframes tick {
  from {
    background-color: var(--tick-color);
  }
  to {
    background-color: var(--background-color);
  }
}

.SlotWheel--tick > .SlotWheel__value {
  will-change: color;
}

.SlotWheel--set {
  background: var(--set-color);
}

.SlotWheel--set > .SlotWheel__value {
  color: var(--white-color);
}

.SlotWheel__value {
  text-align: center;
  font-size: var(--title-size);

  color: var(--type-text-color);
  font-family: monospace;
  position: relative;
  top: calc(var(--caption-size) / 2);
  hyphens: auto;

  /* annoyingly, autoprefixer doesn't do this for us? */
  -webkit-hyphens: auto;

  transition: color 350ms ease-out;
}

.SlotWheel__name {
  font-size: var(--caption-size);
  text-transform: uppercase;

  position: absolute;
  top: 0;
  background: var(--emphasis-background-color);
  width: 100%;
  text-align: center;
  height: calc(2 * var(--caption-size));
  line-height: calc(2 * var(--caption-size));

  color: var(--set-text-color);
}
</style>
