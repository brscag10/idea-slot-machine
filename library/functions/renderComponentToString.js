import Vue from "vue";
import { createRenderer } from "vue-server-renderer";

/**
 * create an html string from a Vue Component for testing purposes
 *
 * @export
 * @param {VueComponent} VueComponent the vue component we wish to render
 * @returns {Promise<string>} the rendered component string
 */
export default async function renderComponentToString(VueComponent, props = {}) {
  const renderer = createRenderer();

  const ComponentWithProps = Vue.extend(VueComponent);
  const vm = new ComponentWithProps(props).$mount();

  return renderer.renderToString(vm);
}
