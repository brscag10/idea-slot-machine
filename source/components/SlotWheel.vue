<template>
  <section
    class="SlotWheel"
    v-bind:class="{ 'SlotWheel--set': hasSpun && !isSpinning, 'SlotWheel--tick': hasTicked }"
    v-bind:style="{ 'animation-duration': `${tickDuration}ms` }"
    @animationend="_clearTick"
  >
    <div class="SlotWheel__value">{{ displayValue }}</div>
    <h2 class="SlotWheel__name">{{ name }}</h2>
  </section>
</template>


<script lang="ts">
import Component from "vue-class-component";
import ShuffleQueue from "../library/ShuffleQueue";
import Vue from "vue";

import config from "configuration/app.config.yml";
import randomIntFactory from "../library/randomIntFactory";

@Component({
  props: {
    name: String,
    options: Array
  }
})
/** 
 * the SlotWheel, once spun, triggers at a regular interval, taking care to not
 * repeat the same thing twice.
 */
export default class SlotWheel extends Vue {
  displayValue: string = "???";
  hasSpun: boolean = false;
  hasTicked: boolean = false;
  isSpinning: boolean = false;
  queue: ShuffleQueue<string> = new ShuffleQueue<string>(this.options);
  tickDuration: number = 0;

  startingSlotVelocityMS: Range<number> = config.startingSlotVelocityMS;
  finalTickInterval: Range<number> = config.finalTickInterval;

  /**
   * randomizes the final tick duration
   * 
   * @returns {number} the duration, in MS
   */
  randomFinalTickInterval: () => number = () => randomIntFactory(
    this.finalTickInterval.max,
    this.finalTickInterval.min
  )();

  /**
   * randomizes the initial tick velocity
   * 
   * @returns {number} the velocity, in MS
   */
  randomStartingVelocity: () => number = () => randomIntFactory(
    this.startingSlotVelocityMS.max,
    this.startingSlotVelocityMS.min
  )();

  /**
   * spin() starts spinning the slot wheel, once called.
   * 
   * @returns {Promise<void>} a promise that resolves once the wheel has stopped spinning
   */
  spin() {
    this.hasSpun = true;
    this.isSpinning = true;

    const startingVelocity = this.randomStartingVelocity();
    const finalTickInterval = this.randomFinalTickInterval();

    let nextTimeout = startingVelocity;

    return new Promise(resolve => {
      this._tick({ finalTickInterval, nextTimeout }, resolve);
    });
  }

  /**
   * _tick uses the configured frictionCoefficient to determine whether to
   * calculate another PhysicsFrame or stop the spinner
   * 
   * @param {function} resolve callback to invoke if this is the last tick on the wheel
   * @returns {void}
   */
  _tick({ finalTickInterval, nextTimeout }: PhysicsFrame, resolve) {
    this.displayValue = this.queue.pick();

    nextTimeout *= config.frictionCoeffecient;

    if (nextTimeout > finalTickInterval) {
      this.isSpinning = false;
      resolve();
    } else {
      this.tickDuration = nextTimeout / 2;
      this.hasTicked = true;

      setTimeout(() => this._tick({ finalTickInterval, nextTimeout }, resolve), nextTimeout);
    }
  }

  /**
   * clears the `hasTicked` property, which preps the SlotWheel for its next animation
   * 
   * @returns {void}
   */
  _clearTick() {
    this.hasTicked = false;
  }
}
</script>


<style scoped>
.SlotWheel {
  height: 10em;
  border: 1px solid var(--black-color);
  padding: var(--vert-padding) var(--horiz-padding);

  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;

  transition: background-color 350ms ease-out;

  will-change: background-color;
}

.SlotWheel--tick {
  animation-name: tick;
}

@keyframes tick {
  from {
    background-color: var(--tick-color);
  }
  to {
    background-color: transparent;
  }
}

.SlotWheel--tick > .SlotWheel__value {
  will-change: color;
}

.SlotWheel--set {
  background: var(--set-color);
}

.SlotWheel--set > .SlotWheel__value {
  color: var(--set-text-color);
}

.SlotWheel__value {
  text-align: center;
  font-size: var(--title-size);

  transition: color 350ms ease-out;
}

.SlotWheel__name {
  font-size: var(--footnote-size);
  text-transform: uppercase;

  color: var(--type-text-color);
}
</style>
