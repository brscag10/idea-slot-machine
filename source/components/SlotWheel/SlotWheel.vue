<template>
  <section
    class="SlotWheel"
    v-bind:class="{ 'SlotWheel--set': isSpinning === false, 'SlotWheel--tick': hasTicked === true }"
    v-bind:style="{ 'animation-duration': tickDuration + 'ms' }"
    @animationend="_clearTick"
  >
    <div class="SlotWheel__value">{{ displayValue }}</div>
    <h2 class="SlotWheel__name">{{ name }}</h2>
  </section>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";

// TODO - imports
import ShuffleQueue from "/library/ShuffleQueue";
import randomIntFactory from "/library/randomIntFactory";

import config from "configuration/app.config.yml";

@Component({
  props: {
    name: String,
    options: Array
  }
})
export default class SlotWheel extends Vue {
  startingSlotVelocityMS = config.startingSlotVelocityMS;
  finalTickInterval = config.finalTickInterval;
  displayValue = "--";
  hasTicked = null;
  isSpinning = null;
  queue = new ShuffleQueue<string>(this.$props.options);
  randomFinalTickInterval = randomIntFactory(
    this.finalTickInterval.max,
    this.finalTickInterval.min
  );
  randomStartingVelocity = randomIntFactory(
    this.startingSlotVelocityMS.max,
    this.startingSlotVelocityMS.min
  );
  tickDuration = null;

  roll() {
    this.$set(this, "isSpinning", true);

    const startingVelocity = this.randomStartingVelocity();
    const finalTickInterval = this.randomFinalTickInterval();

    let nextTimeout = startingVelocity;

    return new Promise(resolve => {
      this._tick({ startingVelocity, finalTickInterval, nextTimeout }, resolve);
    });
  }

  _tick({ nextTimeout, startingVelocity, finalTickInterval }, resolve) {
    this.$set(this, "displayValue", this.$data.queue.pick());

    nextTimeout *= config.frictionCoeffecient;

    if (nextTimeout > finalTickInterval) {
      this.$set(this, "isSpinning", false);
      resolve();
    } else {
      this.$set(this, "tickDuration", nextTimeout / 2);
      this.$set(this, "hasTicked", true);

      setTimeout(this._tick, nextTimeout);
    }
  }

  _clearTick() {
    this.$set(this, "hasTicked", false);
  }
}
</script>

<style scoped>
.SlotWheel {
  height: 10em;
  border: 1px solid var(--black-color);
  padding: var(--vert-padding) var(--horiz-padding);

  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;

  transition: background-color 350ms ease-out;

  will-change: background-color;
}

.SlotWheel--tick {
  animation-name: tick;
}

@keyframes tick {
  from {
    background-color: var(--tick-color);
  }
  to {
    background-color: transparent;
  }
}

.SlotWheel--tick > .SlotWheel__value {
  will-change: color;
}

.SlotWheel--set {
  background: var(--set-color);
}

.SlotWheel--set > .SlotWheel__value {
  color: var(--set-text-color);
}

.SlotWheel__value {
  text-align: center;
  font-size: var(--title-size);

  transition: color 350ms ease-out;
}

.SlotWheel__name {
  font-size: var(--footnote-size);
  text-transform: uppercase;

  color: var(--type-text-color);
}
</style>
