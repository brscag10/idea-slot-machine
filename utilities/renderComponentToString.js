import Vue from "vue";
import { createRenderer } from "vue-server-renderer";

/**
 * create an html string from a Vue Component for testing purposes
 *
 * @export
 * @param {VueComponent} VueComponent the vue component we wish to render
 * @returns {Promise<string>} the rendered component string
 */
export default async function renderComponentToString(VueComponent) {
  const renderer = createRenderer();
  const v = new Vue({
    el: document.createElement("div"),
    render: h => h(VueComponent)
  });

  return new Promise((resolve, reject) => {
    renderer.renderToString(v, (error, result) => {
      if (error) {
        reject(error);
      }

      resolve(result);
    });
  });
}
