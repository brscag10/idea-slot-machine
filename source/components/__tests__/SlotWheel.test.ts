import SlotWheel from "../SlotWheel.vue";
import renderComponentToString from "../../../utilities/renderComponentToString";

test.concurrent("passes a smoke test", async (): Promise<void> => {
  const result = await renderComponentToString(SlotWheel);

  expect(result).toMatchSnapshot();
});
