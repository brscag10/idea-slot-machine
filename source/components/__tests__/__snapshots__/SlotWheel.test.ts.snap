// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes a smoke test 1`] = `"<section data-server-rendered=\\"true\\" class=\\"SlotWheel\\" style=\\"animation-duration:0ms;\\"><div class=\\"SlotWheel__value\\">???</div> <h2 class=\\"SlotWheel__name\\"></h2></section>"`;
